Add Set C-API

jeremyevans opened the pull request:
This should be a minimal C-API needed to deal with Set objects. It
supports creating the sets, checking whether an element is the set,
adding and removing elements, iterating over the elements, clearing
a set, and returning the size of the set.

As some of the function names I wanted to use in the C-API were
already used internally, I had to rename some existing internal functions.

nobu commented on include/ruby/internal/intern/set.h:
```diff
@@ -0,0 +1,111 @@
+VALUE rb_set_new(void);
+
+/**
+ * Identical to rb_set_new(), except it additionally specifies how many elements 
```
```suggestion
 * Identical to rb_set_new(), except it additionally specifies how many elements
```

jeremyevans commented:
> Should `rb_cSet` also be exposed in `include/ruby/internal/globals.h`? Having it exposed will also allow ZJIT's to optimize for Set objects more easily ([PR](https://github.com/ruby/ruby/pull/13743)).

Yes, it should.  I updated this to include that change.

rhenium commented on set.c:
```diff
@@ -1913,6 +1913,56 @@ compat_loader(VALUE self, VALUE a)
+}
+
+VALUE
+rb_set_new_capa(unsigned long capa)
```
Should this take a `size_t`, to match the type used in `rb_set_size()`?

jeremyevans replied:
The underlying function takes the size as `st_index_t`. I didn't want to have the public header depend on `st.h` types. I think either `size_t` or `unsigned long` would work. For the common platforms, `st_index_t` is `unsigned long` (I think it's `unsigned long long` on Windows).

rhenium replied:
I think `rb_set_new_capa()` and `rb_set_size()` should use the same type. Avoiding `st_index_t` here seems like a good idea to me as well.

I think `size_t` would be simpler, since we seem to assume it has the same size as a pointer type (and happens to be compatible with `st_index_t`), whereas `unsigned long` can sometimes be smaller.

jeremyevans replied:
Makes sense to me. When I rebased today, I included that change.

eregon commented on spec/ruby/optional/capi/set_spec.rb:
```diff
@@ -0,0 +1,96 @@
+
+load_extension("set")
+
+describe "C-API Set function" do
```
Thank you for adding these C API specs, that will definitely help implementing those functions on other Rubies.

jeremyevans commented:
> Headsup: I've exposed `rb_cSet` [in `zjit.c`](https://github.com/ruby/ruby/blob/master/zjit.c#L35) for now. So you'll need to remove it when rebasing this PR.

Rebased and updated to remove it from `zjit.c`.

jeremyevans commented on include/ruby/intern.h:
```diff
@@ -51,6 +51,7 @@
 #include "ruby/internal/intern/re.h"
 #include "ruby/internal/intern/ruby.h"
 #include "ruby/internal/intern/select.h"
+#include "ruby/internal/intern/set.h"
```
I agree that it isn't strictly necessary.  However, I'd like the Set C-API to be usable in a similar manner to the Array/Hash C-API, and I don't believe those require separate includes.  If you do not feel that is appropriate, please let me know and I'll submit a PR to switch to `ruby/set.h`
